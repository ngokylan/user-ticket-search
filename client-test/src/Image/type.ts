import { any, string } from 'prop-types';

export type ImageProps = {
  /** Image src */
  src: string;
  /** Image alt */
  alt?: string;
  /** Image custom id attribute */
  id?: string;
  /** Add custom class to class attribute */
  className?: string;
  /** The intrinsic width of the image in pixels. */
  width?: string;
  /** The intrinsic height of the image in pixels */
  height?: string;
  /**
   * A list of one or more strings separated by commas indicating a set of source sizes.
   * Each source size consists of:
   *   1) A media condition. This must be omitted for the last item.
   *   2) A source size value.
   */
  sizes?: string;
  /**
   * A list of one or more strings separated by commas indicating a set of possible
   * image sources for the user agent to use. Each string is composed of:
   *   1) A URL to an image
   *   2) Optionally, whitespace followed by one of:
   */
  srcSet?: string;
  /** The object-fit property of the image */
  objectFit?: 'fill' | 'contain' | 'cover' | 'none' | 'scale-down';
  /** The rendering type modifier of the image */
  type?: 'rounded';
  /** The role HTML attribute for accessibility */
  role?: string;
};
