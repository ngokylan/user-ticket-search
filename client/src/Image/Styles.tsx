import styled from 'styled-components';

import { ImageProps } from '../Image';

export const StyledImage = styled.img<ImageProps>`
  ${({ objectFit }) =>
    objectFit &&
    `
      object-fit: ${objectFit};
    `}
`;
