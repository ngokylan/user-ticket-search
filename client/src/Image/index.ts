export { default } from './Image';
export * from './type';
