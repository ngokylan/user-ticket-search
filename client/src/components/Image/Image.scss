@import '../../_lib/style/index';
.image {
  /* Modifier */
  &--rounded {
    border-radius: 6px;
  }
}
