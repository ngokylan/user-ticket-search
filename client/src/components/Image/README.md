### When To Use
Please refer to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img for more info

### Examples

Image
```js
<Image src="u1.png" />
<Image src="u2.png" />
<Image src="u3.png" />
<Image src="u4.png" />
<Image src="u5.png" />
<Image src="u6.png" />
<Image src="u7.png" />
```
