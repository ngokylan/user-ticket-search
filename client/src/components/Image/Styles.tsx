import styled from 'styled-components';

import { ImageProps } from '.';

export const StyledImage = styled.img<ImageProps>`
  ${({ objectFit }) =>
    objectFit &&
    `
      object-fit: ${objectFit};
    `}
`;
