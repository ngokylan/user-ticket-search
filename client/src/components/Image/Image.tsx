import React from 'react';
import { ImageProps } from './type';
import { getClass } from '../../_lib/helper';
import './Image.scss';
import { StyledImage } from './Styles';

function getSrcSet(src: string, srcSet?: any): string {
  return srcSet;
}

function getSizes(sizes?: any): string {
  return sizes;
}

function Image({
  src,
  id,
  alt,
  className,
  width,
  height,
  srcSet,
  sizes,
  objectFit,
  type,
  role,
}: ImageProps) {
  const roleName = role ? role : 'img';

  return (
    <StyledImage
      id={id}
      alt={alt}
      src={src}
      width={width}
      height={height}
      sizes={getSizes(sizes)}
      srcSet={getSrcSet(src, srcSet)}
      data-testid={`image-${id || 'default'}`}
      objectFit={objectFit}
      role={roleName}
      className={getClass('image', className, { [`${type}`]: type })}
    />
  );
}

export default Image;
