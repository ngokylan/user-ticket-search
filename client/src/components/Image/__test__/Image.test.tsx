import React from 'react';
import { mount } from 'enzyme';
import Image, { ImageProps } from '..';

const defaultProps: ImageProps = undefined;
const imageSrc =
  'https://theme.zdassets.com/theme_assets/1/7eb075fb3220e4a248a73351fc5ad59e3be64586.gif';
const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Image {...props} />);
  const component = wrapper.find('img.image');

  return { props, wrapper, component };
};

describe('Image', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it('should be match with snapshot', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toMatchSnapshot('1');
    });
  });

  describe('props', () => {
    describe('default', () => {
      it.each([
        ['image-default', {}],
        ['image-custom-id', { id: 'custom-id' }],
      ])('should render test ID properly', (expectedTestId, props) => {
        const { component } = mountComponent(props);
        expect(component.prop('data-testid')).toBe(expectedTestId);
      });
    });
  });

  describe('custom', () => {
    it.each([
      ['id', 'custom-id', { id: 'custom-id' }],
      ['src', imageSrc, { src: imageSrc }],
      ['alt', 'alt text', { alt: 'alt text' }],
      ['width', '100', { width: '100' }],
      ['height', '100', { height: '100' }],
      ['srcSet', imageSrc, { srcSet: imageSrc }],
      ['type', 'rounded', { type: 'rounded' }],
      ['objectFit', 'objectFit', { objectFit: 'objectFit' }],
      ['role', 'logo', { role: 'logo' }],
    ])('should render prop %s as %s for Image', (name, expected, props) => {
      const { wrapper } = mountComponent(props);
      const CheckboxInput = wrapper.find('Image');
      expect(CheckboxInput.prop(name)).toBe(expected);
    });
  });
});
