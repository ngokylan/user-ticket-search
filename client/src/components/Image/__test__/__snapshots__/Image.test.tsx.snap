// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image rendering should be match with snapshot: 1 1`] = `
<img
  className="sc-bdfBwQ gjsWpT image"
  data-testid="image-default"
  role="img"
/>
`;
